[FIRRTL][InferResets] Don't abort on foreign types

fabianschuiki opened the pull request:
Make the `InferResets` pass properly ignore foreign types which by definition don't contribute to the resets being inferred. Before this change the pass would abort on various `cast<FIRRTLType>()` calls. These all become `dyn_cast<FIRRTLType>()` now with graceful handling of the non-FIRRTL type case, which makes the pass more robust.

seldridge commented on lib/Dialect/FIRRTL/Transforms/InferResets.cpp:
```diff
@@ -890,10 +888,11 @@ void InferResetsPass::traceResets(FIRRTLType dstType, Value dst, unsigned dstID,
-    if (dstBase.isa<ResetType>() || srcBase.isa<ResetType>()) {
+  auto dstBase = dstType.dyn_cast<FIRRTLBaseType>();
+  auto srcBase = srcType.dyn_cast<FIRRTLBaseType>();
+  if (!dstBase || dstBase.isGround()) {
```
If this fails, this will get dropped into the `llvm_unreachable`, yes? I think the behavior is to just ignore it which is slightly different. Also, can this be reordered to early exit to remove nesting? 😉

fabianschuiki replied:
Yeah arguably that unreachable is not really useful. This should likely just be a check for reset types and corresponding handling of reset types. And if they're not reset types, just don't do anything.

seldridge reviewed (commented):
Looking good.

This could use a test.

fabianschuiki commented:
Thanks for the feedback @seldridge. I added a proper test and also added better early-outs to that `traceResets` function, which makes it a lot more readable and the `llvm_unreachable` is also gone.
